src/ddsrt/src/sync/freertos/sync.c

```diff
@@ -0,0 +1,505 @@
+    case DDS_RETCODE_OUT_OF_RESOURCES:
+      abort();
+    case DDS_RETCODE_TIMEOUT:
+      return false;
+    default:
+      assert(rc == DDS_RETCODE_OK);
+      break;
+  }
+
+  return true;
+}
+
+bool
+ddsrt_cond_waituntil(
+  ddsrt_cond_t *cond,
+  ddsrt_mutex_t *mutex,
+  dds_time_t abstime)
+{
+  dds_retcode_t rc;
+  dds_time_t time;
+  dds_duration_t reltime;
+
+  assert(cond != NULL);
+  assert(mutex != NULL);
+
+  time = dds_time();
+  reltime = (abstime > time ? abstime - time : 0);
+
+  switch ((rc = cond_timedwait(cond, mutex, reltime))) {
+    case DDS_RETCODE_OUT_OF_RESOURCES:
+      abort();
+    case DDS_RETCODE_TIMEOUT:
+      return false;
+    default:
+      assert(rc == DDS_RETCODE_OK);
+      break;
+  }
+
+  return true;
+}
+
+void ddsrt_cond_signal(ddsrt_cond_t *cond)
+{
+  TaskHandle_t task;
+
+  assert(cond != NULL);
+
+  ddsrt_mutex_lock(&cond->mutex);
+  if (cond->tasks.cnt > 0) {
+    task = cond->tasks.buf[0];
+    cond_pop(cond, task);
+    (void)xTaskNotifyGive(task);
+  }
+  ddsrt_mutex_unlock(&cond->mutex);
+}
+
+void ddsrt_cond_broadcast(ddsrt_cond_t *cond)
+{
+  size_t i;
+
+  assert(cond != NULL);
+
+  ddsrt_mutex_lock(&cond->mutex);
+  if (cond->tasks.cnt > 0) {
+    /* Wake up all blocked tasks (in order). */
+    for (i = 0; i < cond->tasks.cnt; i++) {
+      (void)xTaskNotifyGive(cond->tasks.buf[i]);
+    }
+    /* Compact buffer (tasks will re-register). */
+    (void)cond_resize(cond, 0);
+  }
+  ddsrt_mutex_unlock(&cond->mutex);
+}
+
+#define LOCKED_EXCLUSIVE (-1)
+#define UNLOCKED (0)
+#define LOCKED_SHARED (1)
+
+void ddsrt_rwlock_init(ddsrt_rwlock_t *rwlock)
+{
+  assert(rwlock != NULL);
+  ddsrt_mutex_init(&rwlock->mutex);
+  ddsrt_cond_init(&rwlock->cond);
+  rwlock->state = UNLOCKED;
+  rwlock->rdrs = 0;
+  rwlock->wtrs = 0;
+}
+
+void ddsrt_rwlock_destroy(ddsrt_rwlock_t *rwlock)
+{
+  assert(rwlock != NULL);
+  ddsrt_cond_destroy(&rwlock->cond);
+  ddsrt_mutex_destroy(&rwlock->mutex);
+  memset(rwlock, 0, sizeof(*rwlock));
+}
+
+void ddsrt_rwlock_read(ddsrt_rwlock_t *rwlock)
+{
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  while (rwlock->wtrs > 0) {
+    ddsrt_cond_wait(&rwlock->cond, &rwlock->mutex);
+  }
+  assert(rwlock->state == UNLOCKED ||
+         rwlock->state == LOCKED_SHARED);
+  rwlock->rdrs++;
+  rwlock->state = LOCKED_SHARED;
+  ddsrt_mutex_unlock(&rwlock->mutex);
+}
+
+bool ddsrt_rwlock_tryread(ddsrt_rwlock_t *rwlock)
+{
+  bool locked = false;
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  if (rwlock->wtrs == 0) {
+    assert(rwlock->state == UNLOCKED ||
+           rwlock->state == LOCKED_SHARED);
+    rwlock->rdrs++;
+    rwlock->state = LOCKED_SHARED;
+    locked = true;
+  }
+  ddsrt_mutex_unlock(&rwlock->mutex);
+  return locked;
+}
+
+void ddsrt_rwlock_write(ddsrt_rwlock_t *rwlock)
+{
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  while (rwlock->wtrs > 0) {
```

Comment:
So if I have three threads calling ``rwlock_write`` at the same time, one gets through, the other two will block here with ``wtrs==3``. Once the thread that won calls ``unlock``, the other two will see ``wtrs==2`` and so keep waiting forever. They'd need to check ``state`` as well.

I'm not too worried about performance or fairness right now. Better first get it to provide the exclusion without deadlocks. (And  especially if FreeRTOS+POSIX can be used most of the time.)